ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java

```diff
@@ -185,19 +186,28 @@ public SchemaKTable(
           : "";
       throw new KsqlException(errorMsg + additionalMsg);
     }
+
+    final Function<Formats, ExecutionStep<KTableHolder<K>>> stepSupplier =
+        formats -> ExecutionStepFactory.tableSelectKey(
+            contextStacker,
+            sourceTableStep,
+            formats,
+            keyExpression
+        );
+
+    final ExecutionStep<KTableHolder<K>> placeholderStep =
+        stepSupplier.apply(InternalFormats.PlaceholderFormats.of());
+    final LogicalSchema newSchema = resolveSchema(placeholderStep);
+
     final KeyFormat newKeyFormat = forceInternalKeyFormat
         .map(newFmt -> KeyFormat.of(
             newFmt,
-            SerdeFeaturesFactory.buildInternal(FormatFactory.of(newFmt)),
+            SerdeFeaturesFactory.buildInternalKeyFeatures(newSchema, FormatFactory.of(newFmt)),
             keyFormat.getWindowInfo()))
         .orElse(keyFormat);
 
-    final ExecutionStep<KTableHolder<K>> step = ExecutionStepFactory.tableSelectKey(
-        contextStacker,
-        sourceTableStep,
-        InternalFormats.of(newKeyFormat.getFormatInfo(), valueFormat),
-        keyExpression
-    );
+    final ExecutionStep<KTableHolder<K>> step = stepSupplier.apply(
+        InternalFormats.of(newSchema, newKeyFormat.getFormatInfo(), valueFormat));
```

Comment:
I'd like to unify this with the other methods by introducing an appropriate variant of `buildStepWithFormats()` but it'd be very custom since we also need access to `newKeyFormat` since it's needed in the SchemaKTable returned by the method. It's not clear to me that introducing such a custom variant is worth the unification.